doc/developer/design/20230421_stabilize_monotonic_select.md

```diff
@@ -0,0 +1,145 @@
+
+### Improve `refine_single_time_dataflow` LIR-based Refinement (MaterializeInc/materialize#18732)
+
+It is not always the case the `must_consolidate` must be set for every operator coverted to monotonic in a one-shot `SELECT`. For example, in a single-time dataflow, some operators may produce physically monotonic output regardless of whether their input was even logically monotonic, e.g., reductions. Furthermore, some operators will convert single-time, logically monotonic input into physically monotonic input, e.g., operators creating an arrangement. So if the input to a monotonic operator with forced consolidation came from a reduction, then this operator's `must_consolidate` flag could instead be toggled to `false`.
+
+We propose to split the `refine_single_time_dataflow` LIR-based refinement into two sub-actions. The first consists of the present code, which replaces non-monotonic operators with monotonic ones where consolidation is enforced (e.g., `refine_single_time_operator_selection`). The second takes as input a collection of references for the monotonic operators that were replaced in the plan and refines the setting of the `must_consolidate` flag for these `Plan` nodes (e.g., `refine_single_time_consolidation`).
+
+The `must_consolidate` flag refinement can be implemented as a series of iterative child traversals starting from each monotonic `Plan` node replaced in the first sub-action. For a given monotonic `Plan` node, we would like to assert whether its input is already physically monotonic or whether consolidation must be forced in the `Plan` node itself. To support this decision in a single-time context, we note that `Plan` nodes can be classified into: (a) Monotonicity enforcers, e.g., `Plan::Reduce`, which are LIR nodes that enforce arranged output; (b) Monotonicity keepers, e.g., `Plan::FlatMap`, which just maintain the monotonicity property of their input unchanged; or (c) Monotonicity breakers, e.g., `Plan::Negate` or `Plan::Get` on a non-arranged and non-monotonic source, which introduce negative multiplicities, thus destroying physical monotonicity.
+
+To turn off forced consolidation, it must be true that we cannot find a monotonicity breaker prior to a monotonicity enforcer in any path starting at the monotonic `Plan` node towards the sources. Note that since the `Plan` nodes replaced in the first step are outside of recursive contexts, we know that there are no cycles in the subgraphs rooted at each monotonic `Plan` node, which implies that paths starting at the node must eventually meet a source operator. So the property can be checked for each monotonic `Plan` node by an iterative traversal of the children and descendants of the node where a path traversal is not further expanded at monotonicity enforcers and the `must_consolidate` flag can only be set to `false` if the traversal does not meet any monotonicity breaker.
```

Comment:
Given that data flows in the opposite direction this might be counter-intuitive to grasp, you might want to replace it with:

> To turn off forced consolidation at the root of an LIR plan, it must be true that we cannot find a monotonicity breaker after to a monotonicity enforcer in any path from the sources to the root

Reply:
> So the property can be checked for each monotonic `Plan` node

This sounds line an `O(n^2)` approach (you need to check the descendants of each sub-plan), but the bottom-up mechanism of inferring this information lends itself to some dynamic programming and possible reuse of information inferred for of sub-plans.

Reply:
> Given that data flows in the opposite direction this might be counter-intuitive to grasp, you might want to replace it with:
> 
> > To turn off forced consolidation at the root of an LIR plan, it must be true that we cannot find a monotonicity breaker after to a monotonicity enforcer in any path from the sources to the root

Good point; I added your reformulation suggestion to the document and also updated the proposal to be based on an LIR analysis pass implemented `Interpreter`-style.

Reply:
> > So the property can be checked for each monotonic `Plan` node
> 
> This sounds line an `O(n^2)` approach (you need to check the descendants of each sub-plan), but the bottom-up mechanism of inferring this information lends itself to some dynamic programming and possible reuse of information inferred for of sub-plans.

That is precisely why the proposal was hinting at pruning the iterative traversal in monotonicity enforcers (`Plan::Reduce` being one of them, and thus addressing nesting) and not expanding `Plan:Let` definitions. Unfortunately, the latter results in imprecision in the analysis, wherein some opportunities to turn off consolidation would be lost. So I agree with you that a bottom-up scheme is just much better.